### ISSUE TITLE
MobileNet v2 slower than v1 when loading from Frozen GraphDef

### ISSUE THREAD

**Issue by suryaprakaz** [tensorflow/tensorflow#21196]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04.3 LTS
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
Doesn't apply
- **TensorFlow installed from (source or binary)**:
Binary
- **TensorFlow version (use command below)**:
v1.9.0-0-g25c197e023 1.9.0
- **Python version**:
Python3
- **Bazel version (if compiling from source)**:
Doesn't apply
- **GCC/Compiler version (if compiling from source)**:
Doesn't apply
- **CUDA/cuDNN version**:
CUDA: 9.0 ; cuDNN: 7.1.4
- **GPU model and memory**:
GeForce GTX 1080 Ti
- **Exact command to reproduce**:

### Summary
MobileNet v2 is faster than v1 only when loading from the checkpoint format i.e, Variable Ops (meta, index, data) whereas it is slower than v1 when running in the frozen graph format (Const Ops) (.pb)

### Description
I have two TensorFlow trained models that are in the checkpoint format (meta, index, data) namely mobilenetv1_0.75.ckpt and mobilenetv2_0.75_6.ckpt. 

The model definitions are as described in the [MobileNetv1](https://arxiv.org/abs/1704.04861) and [MobileNetv2](https://arxiv.org/pdf/1801.04381.pdf) papers. Both models are trained with a width_multiplier of 0.75. MobileNetv2 has an expansion factor of 6. This means that MAC wise, v2 is better than v1 (v1: 26.5 Mil, v2: 20.6Mil) and is expected to be slightly faster than v1.

To compare how the models actually perform, I evaluated them in two ways
Method 1: After Training - Inference by loading the models from checkpoints
Method 2: During Deployment - Inference by loading the models from frozen GraphDefs ([freeze_graph](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) for converting Variables to Consts)

**Tools Used for Testing:**
- TF's Timeline Trace Tool
- benchmark_model Tool
- Naive Python time module

### Timeline Tool
I used the TensorFlow's [timeline](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/client/timeline.py) tool to view the execution times in the Chrome Trace format. 

**Method 1:** 
MobileNetv1 Trace

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/7610546/43350531-b0ad3f8e-9225-11e8-95da-d21ac7cdc1e6.png">

MobileNetv2 Trace

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/7610546/43350534-bc7a644a-9225-11e8-8032-ec10bddf52a9.png">

Eyeballing it, V2 is only slightly faster than V1; measuring from Conv1 to SoftMax,

v1: 9.73 ms
v2: 8.153 ms

**Method 2:**

MobileNetv1 Trace

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/7610546/43350556-ea35af48-9225-11e8-8f2d-56ccd7e600d3.png">

MobileNet v2 Trace

<img width="1437" alt="image" src="https://user-images.githubusercontent.com/7610546/43350562-f60ebda0-9225-11e8-992d-5c7ae66a5eb4.png">

As you can clearly see, when the models are loaded from a frozen format, v2 is slower than v1. This affects the performance timings for v2 especially while deployment since models are usually exported in the frozen format.

### Time Measurements with Python's time module

Apart from the time traces above, I calculated the FPS by measuring the time between session.run. Even though the frozen models are faster than the checkpoint format, v2 is slower than v1 in the frozen format. Why is that so?

Model | Checkpoint - FPS | Frozen (FPS)
-- | -- | --
MobileNetv1 | 97.6 | 254.86
MobileNetv2 | 159.22 | 185.04

### Benchmark Model Tool Results

**MobileNet_v1_75**

Command Used:
`bazel-bin/tensorflow/tools/benchmark/benchmark_model --graph=MobileNet_V1_75.pb --input_layer="input/Placeholder" --input_layer_shape="1,64,64,3" --input_layer_type="float" --output_layer="output/Softmax_1" --show_run_order=false --show_time=false --show_memory=false --show_summary=true --show_flops=true`

[Node type] | [count] | [avg ms] | [avg %] | [cdf %] | [mem KB] | [times called]
-- | -- | -- | -- | -- | -- | --
Conv2D | 15 | 1.044 | 21.20% | 21.20% | 8658.432 | 15
gpu:Conv2D | 15 | 0.571 | 11.60% | 32.80% | 0 | 35
BiasAdd | 28 | 0.557 | 11.31% | 44.11% | 0 | 28
Add | 27 | 0.493 | 10.01% | 54.12% | 0 | 27
Mul | 27 | 0.48 | 9.75% | 63.87% | 0 | 27
Relu6 | 27 | 0.414 | 8.41% | 72.28% | 0 | 27
DepthwiseConv2dNative | 13 | 0.392 | 7.96% | 80.24% | 540.672 | 13
Const | 113 | 0.37 | 7.51% | 87.75% | 0 | 113
gpu:Mul | 27 | 0.082 | 1.67% | 89.42% | 0 | 27
gpu:Add | 27 | 0.081 | 1.65% | 91.06% | 0 | 27
gpu:DepthwiseConv2dNative | 13 | 0.061 | 1.24% | 92.30% | 0 | 13
NoOp | 1 | 0.061 | 1.24% | 93.54% | 0 | 2
gpu:BiasAdd | 28 | 0.059 | 1.20% | 94.74% | 0 | 28
Transpose | 2 | 0.055 | 1.12% | 95.86% | 49.152 | 2
gpu:Relu6 | 27 | 0.054 | 1.10% | 96.95% | 0 | 27
gpu:MEMCPYHtoD | 1 | 0.043 | 0.87% | 97.83% | 0 | 1
Softmax | 1 | 0.04 | 0.81% | 98.64% | 0.512 | 1
AvgPool | 1 | 0.033 | 0.67% | 99.31% | 3.072 | 1
gpu:Softmax | 1 | 0.009 | 0.18% | 99.49% | 0 | 3
gpu:AvgPool | 1 | 0.006 | 0.12% | 99.61% | 0 | 1
_Arg | 1 | 0.006 | 0.12% | 99.74% | 0 | 1
gpu:Transpose | 1 | 0.005 | 0.10% | 99.84% | 0 | 1
_Retval | 1 | 0.004 | 0.08% | 99.92% | 0 | 1
Reshape | 1 | 0.003 | 0.06% | 99.98% | 0 | 1
gpu:MEMCPYDtoH | 1 | 0.001 | 0.02% | 100.00% | 0 | 1
Total |   | 4.924 |   |   |  

**MobileNet_v2_75**

Command Used:
`bazel-bin/tensorflow/tools/benchmark/benchmark_model --graph=MobileNet_V2_75.pb --input_layer="input/Placeholder" --input_layer_shape="1,64,64,3" --input_layer_type="float" --output_layer="output/Softmax_1" --show_run_order=false --show_time=false --show_memory=false --show_summary=true --show_flops=true`

[Node type] | [count] | [avg ms] | [avg %] | [cdf %] | [mem KB] | [times called]
-- | -- | -- | -- | -- | -- | --
Conv2D | 39 | 2.967 | 37.97% | 37.97% | 7955.2 | 39
Add | 53 | 1.217 | 15.58% | 53.55% | 0 | 53
gpu:Conv2D | 39 | 1.036 | 13.26% | 66.80% | 0 | 93
Relu6 | 36 | 0.554 | 7.09% | 73.89% | 0 | 36
DepthwiseConv2dNative | 17 | 0.408 | 5.22% | 79.11% | 959.232 | 17
Const | 129 | 0.403 | 5.16% | 84.27% | 0 | 129
Mul | 17 | 0.304 | 3.89% | 88.16% | 0 | 17
AddN | 12 | 0.264 | 3.38% | 91.54% | 0 | 12
gpu:Add | 53 | 0.161 | 2.06% | 93.60% | 0 | 53
NoOp | 1 | 0.076 | 0.97% | 94.57% | 0 | 2
gpu:Relu6 | 36 | 0.073 | 0.93% | 95.51% | 0 | 36
gpu:DepthwiseConv2dNative | 17 | 0.07 | 0.90% | 96.40% | 0 | 17
Transpose | 2 | 0.053 | 0.68% | 97.08% | 49.152 | 2
gpu:Mul | 17 | 0.052 | 0.67% | 97.75% | 0 | 17
gpu:MEMCPYHtoD | 1 | 0.044 | 0.56% | 98.31% | 0 | 1
Softmax | 1 | 0.041 | 0.53% | 98.84% | 0.512 | 1
AvgPool | 1 | 0.031 | 0.40% | 99.23% | 3.84 | 1
gpu:AddN | 12 | 0.025 | 0.32% | 99.55% | 0 | 12
gpu:Softmax | 1 | 0.009 | 0.12% | 99.67% | 0 | 3
gpu:AvgPool | 1 | 0.007 | 0.09% | 99.76% | 0 | 1
_Arg | 1 | 0.006 | 0.08% | 99.83% | 0 | 1
gpu:Transpose | 1 | 0.005 | 0.06% | 99.90% | 0 | 1
_Retval | 1 | 0.004 | 0.05% | 99.95% | 0 | 1
Reshape | 1 | 0.003 | 0.04% | 99.99% | 0 | 1
gpu:MEMCPYDtoH | 1 | 0.001 | 0.01% | 100.00% | 0 | 1
Total |   | 7.814 |   |   |  

**Avg Time
v1: 4.924 ms
v2: 7.814 ms**

### Source code / logs

How I load the model with method 1:

```
    def __load_model(self):
        latest_checkpoint = tf.train.latest_checkpoint(self.args.checkpoint_dir)
        if latest_checkpoint:
            print("Loading model checkpoint {} ...\n".format(latest_checkpoint))
            self.saver.restore(self.sess, latest_checkpoint)
            print("Checkpoint loaded\n\n")
        else:
            print("No checkpoints available!\n\n")
```

How I load the model with method 2:

```
def create_graph(checkpoint_path):
   with tf.gfile.FastGFile(checkpoint_path, 'rb') as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())
        _ = tf.import_graph_def(graph_def, name='')
```

I'm interested to know the details on why v2's performance is only half as good as v1 when in the paper it is discussed that v2 is supposed to be 35% faster.

Note:
I've cross checked this with other hardware such as P40, an i5 7th Gen CPU and an alternative TF version (1.5.0) and this pattern is the same.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @derekjchow: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @derekjchow: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @derekjchow: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @derekjchow: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @derekjchow: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Manuscrit** [comment]
This may help, from https://arxiv.org/abs/1903.08469v1 :

"However, MobileNet V2 uses **depthwise separable convolutions which are not directly supported in GPU firmware (the cuDNN library)**. Therefore, MobileNet V2 tends to be slower than ResNet18 in most experimental setups. Note that the same issue disqualifies usage of the DenseNet architecture [12], since it requires efficient convolution over a non-contiguous tensor, which is still not supported in cuDNN."

**Comment by rmothukuru** [comment]
@suryaprakaz,
Sorry for the delayed response. You can consider using [MobileNet_V3](https://arxiv.org/pdf/1905.02244.pdf) which is the most recent and efficient. Please refer the [documentation of MobileNet](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/README.md) for more information. Thanks!
